src/libraries/Microsoft.PowerFx.Interpreter/Functions/LibraryRegEx.cs

```diff
@@ -0,0 +1,343 @@
+                    fValid = false;
+                    errors.EnsureError(DocumentErrorSeverity.Severe, args[0], TexlStrings.ErrStringExpected);
+                }
+
+                if (argTypes[1].Kind != DKind.String && argTypes[1].Kind != DKind.OptionSetValue)
+                {
+                    fValid = false;
+                    errors.EnsureError(DocumentErrorSeverity.Severe, args[0], TexlStrings.ErrStringExpected);
+                }
+
+                if (argTypes.Length == 3)
+                {
+                    if (argTypes[0].Kind != DKind.String && argTypes[0].Kind != DKind.ObjNull)
+                    {
+                        fValid = false;
+                        errors.EnsureError(DocumentErrorSeverity.Severe, args[3], TexlStrings.ErrStringExpected);
+                    }
+                }
+
+                return fValid;
+            }
+
+            public virtual async Task<FormulaValue> InvokeAsync(FormulaValue[] args, CancellationToken cancellationToken)
+            {
+                if (args[0] is BlankValue)
+                {
+                    return new BooleanValue(IRContext.NotInSource(FormulaType.Boolean), false);
+                }
+
+                string inputString = ((StringValue)args[0]).Value;
+                string regularExpression = ((StringValue)args[1]).Value;
+                string matchOptions = args.Length == 3 ? ((StringValue)args[2]).Value : DefaultMatchOption;
+                RegexOptions regOptions = RegexOptions.None;
+
+                if (!matchOptions.Contains("c"))
+                {
+                    return FormulaValue.New(false);
+                }
+
+                if (matchOptions.Contains("i"))
+                {
+                    regOptions |= RegexOptions.IgnoreCase;
+                }
+
+                if (matchOptions.Contains("m"))
+                {
+                    regOptions |= RegexOptions.Multiline;
+                }
+
+                if (matchOptions.Contains("^") && !regularExpression.StartsWith("^", StringComparison.Ordinal))
+                {
+                    regularExpression = "^" + regularExpression;
+                }
+
+                if (matchOptions.Contains("$") && !regularExpression.EndsWith("$", StringComparison.Ordinal))
+                {
+                    regularExpression += "$";
+                }
+
+                try
+                {
+                    return RegExFunction(inputString, regularExpression, regOptions);
+                }
+                catch (RegexMatchTimeoutException rexTimeoutEx)
+                {
+                    return new ErrorValue(args[0].IRContext, new ExpressionError()
+                    {
+                        Message = $"Regular expression timeout (above {rexTimeoutEx.MatchTimeout.TotalMilliseconds} ms) - {rexTimeoutEx.Message}",
+                        Span = args[0].IRContext.SourceContext,
+                        Kind = ErrorKind.QuotaExceeded
```

Comment:
I think we should have a timeout error, separate from Quota.  The description in https://learn.microsoft.com/en-us/power-platform/power-fx/error-handling says it is about storage.  It doesn't need to be specific to regular expressions with a generic message like "Operation timed out."  @CarlosFigueiraMSFT, your thoughts?

Reply:
I have created ErrorKind.Timeout